File: ios/PazChurch/Features/Agenda/AgendaListViewModel.swift
import Observation
import Shared
import SwiftUI

@MainActor
@Observable
class AgendaListViewModel {
    var events: [AgendaEvent] = []
    var isLoading = true
    var isLoadingMore = false
    var hasReachedEnd = false

    private let agendaRepository: AgendaRepository
    private var currentPage: Int32 = 1
    private var isRequestInFlight = false

    private static let pageSize: Int32 = 20

    init(agendaRepository: AgendaRepository) {
        self.agendaRepository = agendaRepository
    }

    func loadFirstPage() async {
        currentPage = 1
        isLoading = true
        hasReachedEnd = false
        isRequestInFlight = true
        defer { isRequestInFlight = false }
        do {
            let result = try await agendaRepository.getEvents(page: 1, limit: Self.pageSize)
            events = result
            hasReachedEnd = result.count < Self.pageSize
            isLoading = false
        } catch {
            print("[AgendaListVM] loadFirstPage FAILED — \(type(of: error)): \(error)")
            isLoading = false
        }
    }

    func loadMore() {
        guard !isRequestInFlight, !isLoading, !isLoadingMore, !hasReachedEnd else { return }
        isRequestInFlight = true
        Task {
            isLoadingMore = true
            defer {
                isLoadingMore = false
                isRequestInFlight = false
            }
            let nextPage = currentPage + 1
            do {
                let newEvents = try await agendaRepository.getEvents(page: nextPage, limit: Self.pageSize)
                currentPage = nextPage
                events += newEvents
                hasReachedEnd = newEvents.count < Self.pageSize
            } catch {
                // silently keep existing events; user can scroll back up and retry
            }
        }
    }
}

File: kmp-mobile/android/src/main/kotlin/br/church/paz/android/ui/features/agenda/AgendaListViewModel.kt
package br.church.paz.android.ui.features.agenda

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.church.paz.shared.domain.model.AgendaEvent
import br.church.paz.shared.domain.repository.AgendaRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class AgendaListUiState(
    val events: List<AgendaEvent> = emptyList(),
    val isLoading: Boolean = true,
    val isRefreshing: Boolean = false,
    val isLoadingMore: Boolean = false,
    val error: String? = null,
    val hasReachedEnd: Boolean = false,
)

private const val PAGE_SIZE = 20

class AgendaListViewModel(
    private val agendaRepository: AgendaRepository,
) : ViewModel() {
    private val _uiState = MutableStateFlow(AgendaListUiState())
    val uiState: StateFlow<AgendaListUiState> = _uiState.asStateFlow()

    private var currentPage = 1
    private var isRequestInFlight = false

    init {
        loadFirstPage(showSkeleton = true)
    }

    /** Pull-to-refresh entry point — re-invokes the same load path without the full-screen skeleton. */
    fun refresh() = loadFirstPage(showSkeleton = false)

    private fun loadFirstPage(showSkeleton: Boolean) {
        currentPage = 1
        viewModelScope.launch {
            isRequestInFlight = true
            _uiState.value =
                _uiState.value.copy(
                    isLoading = showSkeleton,
                    isRefreshing = !showSkeleton,
                )
            runCatching { agendaRepository.getEvents(page = 1, limit = PAGE_SIZE) }
                .onSuccess { events ->
                    _uiState.value =
                        AgendaListUiState(
                            events = events,
                            isLoading = false,
                            isRefreshing = false,
                            hasReachedEnd = events.size < PAGE_SIZE,
                        )
                }.onFailure {
                    _uiState.value =
                        _uiState.value.copy(
                            isLoading = false,
                            isRefreshing = false,
                            error = it.message,
                        )
                }
            isRequestInFlight = false
        }
    }

    fun loadMore() {
        val state = _uiState.value
        if (isRequestInFlight || state.isLoading || state.isLoadingMore || state.hasReachedEnd) return
        isRequestInFlight = true
        viewModelScope.launch {
            val nextPage = currentPage + 1
            _uiState.value = state.copy(isLoadingMore = true)
            runCatching { agendaRepository.getEvents(page = nextPage, limit = PAGE_SIZE) }
                .onSuccess { newEvents ->
                    currentPage = nextPage
                    _uiState.value =
                        _uiState.value.copy(
                            events = _uiState.value.events + newEvents,
                            isLoadingMore = false,
                            hasReachedEnd = newEvents.size < PAGE_SIZE,
                        )
                }.onFailure {
                    _uiState.value = _uiState.value.copy(isLoadingMore = false)
                }
            isRequestInFlight = false
        }
    }

    fun retry() = loadFirstPage(showSkeleton = true)
}
